### PR TITLE
Improving the italic and bold styles

### LINKED ISSUE #84: Unexpected italic style in github.com diff views
If in the markdown code there is a `*`, everything after that character will have italic style.
I saw this recently in the GitHub diffs (e.g. [this one](https://github.com/github/git-lfs/commit/d8f780329b64e789553bc8ccccfb993ebc430325)):

screenshot from 2015-04-09 13 56 03
screenshot from 2015-04-09 14 00 08

### PR DESCRIPTION
**Pull request by IonicaBizau**
This will fix #84. Since I'm not a regex guru I got some inspiration from the [`marked` package](https://github.com/chjj/marked/blob/2b5802f258c5e23e48366f2377fbb4c807f47658/lib/marked.js#L448-L462).

 - [x] Italic style fix
 - [x] Bold style fix
 - [x] Combined (?)
 - [ ] Handle styles on multiple lines. Example:

   ```md
   *Lorem
   ipsum
   dolor
   sit
   amet*
   ```

## Before
image

### After

image

Waiting for feedback if the direction is good, then I will continue. :smile: 

/cc @izuzak

### CODE REVIEW

**Comment by mnquintana** [comment]
This is a good start! Eventually you'll want to add specs for this.

(Also, [:lipstick: is deprecated in favor of :art:](https://github.com/atom/atom/blob/master/CONTRIBUTING.md#git-commit-messages) for code structure / formatting changes, and this isn't really a code structure / formatting change – it's really a :bug: )

**Comment by IonicaBizau** [comment]
Indeed, the *art* is much better than a *lipstick*. :smile: 

OK! Will make sure that the tests are passing and I will also try to add the specs for this.

**Comment by tpoisot** [comment]
Really cool. I think that having the italic/bold marks with a different syntax style than the italic/bold content is useful.

**Comment by IonicaBizau** [comment]
I fixed the bold and bold + italic styles as well:

image

Is it possible to pass the regex flags in this format (would need to pass `m` for multiline)? Or how to fix the multiline thing (the last two lines from the screenshot are supposed to have bold style)?

Also, I'm not sure where to start fixing the failing tests. :green_heart:

**Comment by tpoisot** [comment]
I think you should capture the `*` and `_` too, so that they can be separately styled. 

And the tests are in the `spec/`folder

**Comment by IonicaBizau** [comment]
> I think you should capture the * and _ too, so that they can be separately styled.

That's already done, but not separately, but I was asking how can I apply **bold style** on the last two rows? I guess it's related to the `m` (multiline) regex flag.

I will try to fix the failing tests and then writing the new specs on this issue.

**Comment by IonicaBizau** [comment]
Is it possible to log only the first failing test when running `apm test`?

**Comment by izuzak** [comment]
> Is it possible to log only the first failing test when running apm test?

I don't know if that's possible (see https://github.com/jasmine/jasmine/issues/414), but you can focus on a specific `describe` or `it` block by prefixing it with "f". See https://github.com/atom/jasmine-focused#using. Does that help perhaps?

**Comment by IonicaBizau** [comment]
Yes, that's a good solution. Thanks! :+1: 

Also, is it possible to merge repetitive snippets like the following ones, into one?

```cson
{
  'match': '...',
  'captures': 
    '2': 
       'name': 'markup.bold.italic.gfm'
    '3': 
       'name': 'markup.bold.italic.gfm'
    '4': 
       'name': 'markup.bold.italic.gfm'
}
```

...maybe something like this:

```cson
{
  'match': '...',
  'captures': 
    '2': 
    '3': 
    '4': 
       'name': 'markup.bold.italic.gfm'
}
```

I'm not very familiar with the cson format.

**Comment by izuzak** [comment]
I don't think that's possible.

File: grammars/gfm.cson [hunk 1]
```diff
@@ -15,34 +15,28 @@
...
-  {
-    'begin': '(?<=^|[^\\w\\d\\*])\\*\\*(?!$|\\*|\\s)'
-    'end': '(?<!^|\\s)\\*\\**\\*(?=$|[^\\w|\\d])'
-    'name': 'markup.bold.gfm'
-  }
-  {
-    'begin': '(?<=^|[^\\w\\d_])__(?!$|_|\\s)'
-    'end': '(?<!^|\\s)__*_(?=$|[^\\w|\\d])'
-    'name': 'markup.bold.gfm'
+    'match': '^___([\\s\\S]+?)___(?!_)|^\\*\\*\\*([\\s\\S]+?)\\*\\*\\*(?!\\*)'
```
Comment (IonicaBizau) [inline grammars/gfm.cson]: I currently have `'match': '(?<=^|[^\\w\\d\\*])(\\*\\*\\*)([\\s\\S]+?)(\\*\\*\\*)(?=$|[^\\w|\\d])'`, and I [the tokens are](https://github.com/atom/language-gfm/blob/79b391a22d17da19cd7db4e475cfa1147157ebab/spec/gfm-spec.coffee#L38):

```json
[
    {
        "value": "this is ",
        "scopes": [
            "source.gfm"
        ],
        "bufferDelta": 8,
        "hasPairedCharacter": false,
        "screenDelta": 8
    },
    {
        "value": "***bold italic***",
        "scopes": [
            "source.gfm",
            "markup.bold.italic.gfm"
        ],
        "bufferDelta": 17,
        "hasPairedCharacter": false,
        "screenDelta": 17
    },
    {
        "value": " text",
        "scopes": [
            "source.gfm"
        ],
        "bufferDelta": 5,
        "hasPairedCharacter": false,
        "screenDelta": 5
    }
]
```

I expected that by doing `(\\*\\*\\*)` these will be grouped in another element (`{ value: "***" }`). The regex101 tester works fine in this direction:

image

How can I catch these characters into a separate group when using `apm test`? Also, what tool is used for parsing this regex?

**Comment by leipert** [comment]
The main problem I found: mixed emphasis are handled badly, as well as some cases of escaped `*` and `_`.

I created this test markdown, below you can see how github renders it. As you can see some github rendered stuff is not consistent.

```
### Emphasis

#### Bold `_`

+ __a__
+ __ab__
+ __abc__
+ __bbc___
+ __x + y + z__

NOT:

+ as __danach__fasd
+ as __danach____fasd
+ davor__asd__ as
+ in__a__word
+ in 1__8__99 numbers
+ __ abc __
+ __ abc__
+ __abc __

MULTILINE:
__sadsad
asdsadasd
__

#### Italic `_`

+ _a_
+ _ab_
+ _abc_
+ _bbc__
+ _x + y + z_

NOT:

+ as _danach_fasd
+ as _danach___fasd
+ davor_asd_ as
+ in_a_word
+ in 1_8_99 numbers
+ _ abc _
+ _ abc_
+ _abc _

#### Bold `*`

+ **a**
+ **ab**
+ **\\**
+ in**a**word
+ in 1**8**99 numbers
+ as **danach**fasd
+ davor**asd** as
+ **abc**
+ **bbc***
+ **x + y + z**

NOT:

+ **\**
+ ** **, *****, **abc*\*
+ ** abc **
+ ** abc**
+ **abc **

#### Italic `*`

+ *a*
+ *ab*
+ *\\*
+ in*a*word
+ as *danach*fasd
+ davor*asd* as
+ in 1*8*99 numbers
+ *abc*
+ *bbc**
+ *x + y + z*

NOT:

+ *\*, * *, ***, *abc*
+ * abc *
+ * abc*
+ *abc *

#### Mixed

Works:

this is ***bold italic*** text

***BoldItalic***
___BoldItalic___
**Bold *BoldItalic* Bold**
__Bold *BoldItalic* Bold__
_Italic **ItalicBold** Italic_
**Bold _BoldItalic_ Bold**
*Italic __ItalicBold__ Italic*
__Bold _BoldItalic_ Bold__

Works not:

*Italic **ItalicBold** Italic*

_Italic __ItalicBold__ Italic_



Should not work:

_\_, _ _, ____
__\__, __ __, _____

__asd \*neither is this\* asd__

**asd \_neither is this\_ asd**

__asd \**neither is this*\* asd__

__asd \_neither is this\_ asd__

#### Rest

This text is _emphasized with underscores_, and this
is *emphasized with asterisks*

This is **strong emphasis** and __with underscores__.

Working:
***bold italic***

Not Working:
*italic **bold italic** italic*
**bold *bold italic* bold**

This is * not emphasized *, and \*neither is this\*.
```

### Emphasis

#### Bold `_`

+ __a__
+ __ab__
+ __abc__
+ __bbc___
+ __x + y + z__

NOT:

+ as __danach__fasd
+ as __danach____fasd
+ davor__asd__ as
+ in__a__word
+ in 1__8__99 numbers
+ __ abc __
+ __ abc__
+ __abc __

MULTILINE:
__sadsad
asdsadasd
__

#### Italic `_`

+ _a_
+ _ab_
+ _abc_
+ _bbc__
+ _x + y + z_

NOT:

+ as _danach_fasd
+ as _danach___fasd
+ davor_asd_ as
+ in_a_word
+ in 1_8_99 numbers
+ _ abc _
+ _ abc_
+ _abc _

#### Bold `*`

+ **a**
+ **ab**
+ **\\**
+ in**a**word
+ in 1**8**99 numbers
+ as **danach**fasd
+ davor**asd** as
+ **abc**
+ **bbc***
+ **x + y + z**

NOT:

+ **\**
+ ** **, *****, **abc*\*
+ ** abc **
+ ** abc**
+ **abc **

#### Italic `*`

+ *a*
+ *ab*
+ *\\*
+ in*a*word
+ as *danach*fasd
+ davor*asd* as
+ in 1*8*99 numbers
+ *abc*
+ *bbc**
+ *x + y + z*

NOT:

+ *\*, * *, ***, *abc*
+ * abc *
+ * abc*
+ *abc *

#### Mixed

Works:

this is ***bold italic*** text

***BoldItalic***
___BoldItalic___
**Bold *BoldItalic* Bold**
__Bold *BoldItalic* Bold__
_Italic **ItalicBold** Italic_
**Bold _BoldItalic_ Bold**
*Italic __ItalicBold__ Italic*
__Bold _BoldItalic_ Bold__

Works not:

*Italic **ItalicBold** Italic*

_Italic __ItalicBold__ Italic_



Should not work:

_\_, _ _, ____
__\__, __ __, _____

__asd \*neither is this\* asd__

**asd \_neither is this\_ asd**

__asd \**neither is this*\* asd__

__asd \_neither is this\_ asd__

#### Rest

This text is _emphasized with underscores_, and this
is *emphasized with asterisks*

This is **strong emphasis** and __with underscores__.

Working:
***bold italic***

Not Working:
*italic **bold italic** italic*
**bold *bold italic* bold**

This is * not emphasized *, and \*neither is this\*.

**Comment by leipert** [comment]
Please have a look at my comment in #44. There is simply no satisfying way to implement multiline emphasized/bold text with the current grammar, as far as I can see it.

**Comment by burodepeper** [comment]
Hey @leipert (and others),

I've been looking into this as well (see #120, different thing, but got me thinking about Markdown), and I think multiline emphasis is just not part of the philosophy behind Markdown. From that same philosophy, I believe headings, for example, shouldn't be able to contain inline markup. Markdown is about simplicity. If you want complex markup, use HTML.

I haven't looked at all your examples above, but how about we write specs for how we believe the simple version of Markup should be implemented, and ignore the weird edge cases?

**Comment by leipert** [comment]
@burodepeper You are right, markdown is all about simplicity, but it is also really expressive. That it is the reason why you are able to compile it to nice looking HTML, PDF or even Powerpoint with tools like pandoc.

Even the original MD definition by John Gruber is allowing you advanced stuff like inline HTML https://daringfireball.net/projects/markdown/syntax. The original spec does not forbid you to use something like this:

```markdown
# This is a *fancy* header
```

This is the reason why most markdown renderers allow you to write inline emphasis.

Furthermore the syntax highlighting should reflect the actual render. It makes no sense, that the language-gfm package shows you another syntax than the markdown-preview output.

On a side note:
After dealing several hours (something like 80) with Atoms rendering while creating language-pfm, I came to several conclusions:

1. Syntax highlighting with Regex sucks. Especially markdown with Atoms Grammar Regex Parsers, as markdown syntax heavily relies on empty lines, previous lines and new paragraphs, which cannot be detected with the current Grammar Parser.
2. There are too many different markdown flavors. I hope, that http://spec.commonmark.org/ will overcome this in the future.
3. The only reliable syntax highlighting of markdown can be done with help of Syntax Trees. Hopefully this is somewhat possible in the future, as markdown is not really made for line by line parsing.

**Comment by burodepeper** [comment]
> Furthermore the syntax highlighting should reflect the actual render. It makes no sense, that the language-gfm package shows you another syntax than the markdown-preview output.

I don't agree on that. I believe Markdown is about nothing more than semantics (as html is supposed to be). Aren't you supposed to be able to semantically read a Markdown document without any syntax highlighting? I've started using Markdown in my emails, and people understand marking words with asterisks, and specifying headers and `<hr>`s.

> Syntax highlighting with Regex sucks.

That's something I _do_ agree on! ; )

> I hope, that http://spec.commonmark.org/ will overcome this in the future.

I've heard about it, but haven't really looked into it. Putting it on my list right now.

---

In short, I use markdown files a lot, so I don't mind spending time to improve the language-gfm package. And since it's _Github Flavored_, who's in charge of deciding the level of simplicity?

**Comment by leipert** [comment]
I also use markdown files a lot, which brought me to participate in these three packages:
*   [markdown-preview-plus](https://atom.io/packages/markdown-preview-plus). An awesome Markdown Preview (new release coming tonight)
*   [language-pfm](https://atom.io/packages/language-pfm). Pandoc Flavored Markdown, which is a fork of this package and works out a lot of the quirks, while maintaing compability to this package.
*   [linter-markdown](https://atom.io/packages/linter-markdown). Markdown Linter

**Comment by burodepeper** [comment]
Thanks, I'm going to take a look at them.

**Comment by tpoisot** [comment]
@burodepeper I don't think your point on nested emphasis or emphasis in headers not being within the markdown spirit is right. In fact, if you look at the original markdown [implementation](http://daringfireball.net/projects/markdown/dingus), `# a *b*` is parsed as `<h1>a <em>b</em></h1>`. And emphasis *is* semantics!

I would also be happy if [CommonMark](http://spec.commonmark.org/0.21/) gained more prevalence. Github flavored markdown is already more or less CommonMark compliant, by the way.

**Comment by IonicaBizau** [comment]
I'm going to close this since I just feel I'm not ready to understand this regex black magic yet. :smile: :crystal_ball:
